diff -ur linux-4.9-original/include/linux/compiler.h linux-4.9/include/linux/compiler.h
--- linux-4.9-original/include/linux/compiler.h	2016-12-11 19:17:54.000000000 +0000
+++ linux-4.9/include/linux/compiler.h	2017-02-28 16:21:25.432490539 +0000
@@ -132,12 +132,8 @@
  * value is always the same.  This idea is taken from a similar patch
  * written by Daniel Walker.
  */
-# ifndef likely
-#  define likely(x)	(__builtin_constant_p(x) ? !!(x) : __branch_check__(x, 1))
-# endif
-# ifndef unlikely
-#  define unlikely(x)	(__builtin_constant_p(x) ? !!(x) : __branch_check__(x, 0))
-# endif
+#define likely(x) (x)
+#define unlikely(x) (x)

 #ifdef CONFIG_PROFILE_ALL_BRANCHES
 /*
@@ -164,8 +160,8 @@
 #endif /* CONFIG_PROFILE_ALL_BRANCHES */

 #else
-# define likely(x)	__builtin_expect(!!(x), 1)
-# define unlikely(x)	__builtin_expect(!!(x), 0)
+# define likely(x)	(x)
+# define unlikely(x)	(x)
 #endif

 /* Optimization barrier */
diff -ur linux-4.9-original/Makefile linux-4.9/Makefile
--- linux-4.9-original/Makefile	2016-12-11 19:17:54.000000000 +0000
+++ linux-4.9/Makefile	2017-02-28 16:22:23.956498656 +0000
@@ -797,10 +797,10 @@
 KBUILD_ARFLAGS := $(call ar-option,D)

 # check for 'asm goto'
-ifeq ($(shell $(CONFIG_SHELL) $(srctree)/scripts/gcc-goto.sh $(CC)), y)
-	KBUILD_CFLAGS += -DCC_HAVE_ASM_GOTO
-	KBUILD_AFLAGS += -DCC_HAVE_ASM_GOTO
-endif
+#ifeq ($(shell $(CONFIG_SHELL) $(srctree)/scripts/gcc-goto.sh $(CC)), y)
+#	KBUILD_CFLAGS += -DCC_HAVE_ASM_GOTO
+#	KBUILD_AFLAGS += -DCC_HAVE_ASM_GOTO
+#endif

 include scripts/Makefile.kasan
 include scripts/Makefile.extrawarn
