diff --git a/arch/arm/mm/mmu.c b/arch/arm/mm/mmu.c
index f5fdd5c..6665800 100644
--- a/arch/arm/mm/mmu.c
+++ b/arch/arm/mm/mmu.c
@@ -466,8 +466,8 @@ static void __init build_mem_type_table(void)
 	 * v6/v7 kernels), so we must use a separate memory type for user
 	 * r/o, kernel r/w to map the vectors page.
 	 */
-	if (cpu_arch == CPU_ARCH_ARMv6)
-		vecs_pgprot |= L_PTE_MT_VECTORS;
+//	if (cpu_arch == CPU_ARCH_ARMv6)
+//		vecs_pgprot |= L_PTE_MT_VECTORS;
 
 	/*
 	 * Check is it with support for the PXN bit
